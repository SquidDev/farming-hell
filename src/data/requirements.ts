import type { Id, OwnedServant, Servant, Target, UpgradeRequirements } from ".";
import type { Store } from "../store";
import { getAscensionTargetFast } from "./ascension";
import { expEffective, expNormal, grailId } from "./constants";

export type Requirements = {
  qp: number,
  exp: number,
  expBonus: number,
  items: Map<Id<"item">, number>,
}

const addRequirements = (
  requirements: Requirements,
  { current, target }: Target, offset: number,
  materials: Array<UpgradeRequirements>,
): void => {
  if (target === undefined) return;
  if (current === undefined) current = offset;

  for (let i = current; i < target; i++) {
    const thisLevel = materials[i - offset];
    for (const item of thisLevel.items) {
      requirements.items.set(item.item, (requirements.items.get(item.item) ?? 0) + item.amount);
    }

    requirements.qp += thisLevel.qp;
  }
};

const addExpRequirements = (requirements: Requirements, store: Store, servant: OwnedServant, details: Servant): void => {
  let current = servant.level.current;
  const target = servant.level.target;
  if (current === undefined) current = 1;
  if (target === undefined || current >= target) return;

  const currentExp = store.expGrowth[current - 1];
  const targetExp = store.expGrowth[target - 1];
  if (currentExp === undefined || targetExp === undefined) return;

  const cost = targetExp - currentExp;
  requirements.exp += Math.ceil(cost / expNormal);
  requirements.expBonus += Math.ceil(cost / expEffective);

  const maxLevel = details.lvlMax[4];
  if (target > maxLevel) {
    let grails = 0;
    for (const grail of store.grailCosts[details.rarity]) {
      requirements.qp += grail.qp;
      grails++;
      if (maxLevel + grail.add >= target) break;
    }

    requirements.items.set(grailId, (requirements.items.get(grailId) ?? 0) + grails);
  }
};

export const addSplitServantRequirements = (
  ascensionReqs: Requirements, skillReqs: Requirements,
  store: Store, servant: OwnedServant, details: Servant
): void => {
  for (const skill of servant.skills) addRequirements(skillReqs, skill, 1, details.skillMaterials);
  if (details.ascensionMaterials) {
    addRequirements(ascensionReqs, getAscensionTargetFast(details, servant), 0, details.ascensionMaterials);
  }

  addExpRequirements(ascensionReqs, store, servant, details);
};

export const addServantRequirements = (requirements: Requirements, store: Store, servant: OwnedServant, details: Servant): void =>
  addSplitServantRequirements(requirements, requirements, store, servant, details);

export const newRequirements = (): Requirements => ({ qp: 0, exp: 0, expBonus: 0, items: new Map() });
