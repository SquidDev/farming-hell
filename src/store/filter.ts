import { type Filters, type OwnedServant, Priority, type Target } from "../data";

export const defaultFilters: Filters = Object.freeze({
  maxed: true,
  priority: Object.freeze({
    [Priority.High]: true,
    [Priority.Medium]: true,
    [Priority.Low]: true,
    [Priority.Unsummoned]: true,
  }),
});

const isMaxed = ({ current, target }: Target): boolean => current !== undefined && target !== undefined && current === target;
const isMaxedLax = ({ current, target }: Target): boolean => current === target;

export const matchesFilter = (filters: Filters, servant: OwnedServant): boolean => {
  if (filters.priority[servant.priority] === false) return false;

  return filters.maxed || !isMaxed(servant.level) || !servant.skills.every(isMaxed) || !servant.appendSkills.every(isMaxedLax);
};
