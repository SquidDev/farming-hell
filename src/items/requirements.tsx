import { useContext, createContext, type FunctionComponent, type ReactElement } from "react";
import { observer } from "mobx-react-lite";

import { byPriority } from "../data/order";
import { type Requirements, addServantRequirements, newRequirements } from "../data/requirements";
import type { Store } from "../store";
import type { OwnedServant } from "../data";
import { EmberWithAmount, ItemWithAmount, QPWithAmount } from "./item";

export const ShowDifference = createContext(false);

/** A display for a set of requirements, including QP and items. */
export const ShortRequirements: FunctionComponent<{ requirements: Requirements, store: Store }> = observer(function ShortRequirements({ requirements: { items, qp, exp, expBonus }, store }) {
  const showDifference = useContext(ShowDifference);
  const ownedItems = store.ownedItems;

  const itemList: Array<ReactElement<unknown>> = [...items.keys()].sort(byPriority(store.itemLookup)).map(key => {
    let amount = items.get(key)!;
    if (showDifference) amount = (ownedItems.get(key) || 0) - amount;
    return <ItemWithAmount key={key} item={key} store={store} amount={amount} difference={showDifference} />;
  });
  return <>
    {qp > 0 || exp > 0 ? <div className="flex flex-row justify-between">
      <QPWithAmount className="mb-2" store={store} amount={qp} />
      <EmberWithAmount className="mb-2" store={store} amount={exp} bonusAmount={expBonus} />
    </div> : null}
    {itemList}
  </>;
});

/** Convenience wrapper over `ShortRequirements` for a servant's requirements.  */
export const ShortServantRequirements: FunctionComponent<{ servant: OwnedServant, store: Store }> = observer(function ShortServantRequirements({ servant, store }) {
  const details = store.servantLookup.get(servant.id);
  const requirements = newRequirements();
  if (details) addServantRequirements(requirements, store, servant, details);

  return <ShortRequirements requirements={requirements} store={store} />;
});
