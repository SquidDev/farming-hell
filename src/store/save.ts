import { runInAction, toJS } from "mobx";

import save from "../support/save";
import type { Store } from ".";
import { Filters, Id, OwnedServant, Target, makeId } from "../data";
import { defaultFilters } from "./filter";
import { qpId } from "../data/constants";

type LatestSchema = {
  version: 1,
  servants: Array<OwnedServant>,
  items?: Array<[Id<"item">, number]>,
  filters?: Filters,
};

type Schemas
  = Array<OwnedServant>
  | LatestSchema;

const load = (store: Store, data: Schemas): void => {
  let servants: Array<OwnedServant> = [];
  const items: Map<Id<"item">, number> = new Map();
  let filters: Filters = defaultFilters;

  if (data instanceof Array) {
    servants = data;
  } else if (data.version === 1) {
    servants = data.servants;
    if (data.items) {
      for (const [id, count] of data.items) items.set(id, count);
    }
    if (data.filters) filters = data.filters;
  } else {
    throw new Error("Failed to parse this data.");
  }

  // Convert our legacy QP item (id=5) to the current one (id=1).
  const oldQpId = makeId("item", 5);
  const qp = items.get(oldQpId);
  if (qp !== undefined) {
    items.set(qpId, qp);
    items.delete(oldQpId);
  }

  runInAction(() => {
    store.ownedServants = servants;
    store.ownedItems = items;
    store.filters = filters;
  });
};

export const loadFromLocalStorage = (store: Store): void => {
  const servantData = window.localStorage.getItem("servants");
  if (!servantData) return;

  load(store, JSON.parse(servantData) as Schemas);
};

export const saveToLocalStorage = (store: Store): void => {
  const data: LatestSchema = {
    version: 1,
    servants: toJS(store.ownedServants),
    items: [...store.ownedItems.entries()],
    filters: store.filters,
  };
  window.localStorage.setItem("servants", JSON.stringify(data));
};

export const loadFromJson = (store: Store, blob: Blob): Promise<void> => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.addEventListener("load", () => {
      try {
        load(store, JSON.parse(reader.result as string) as Schemas);
      } catch (e) {
        reject(`Failed to read file (${e})`); // eslint-disable-line @typescript-eslint/restrict-template-expressions
        return;
      }

      resolve();
    });
    reader.addEventListener("error", () => reject("Failed to read file"));
    reader.readAsText(blob);
  });
};

export const pickAndLoadFromJson = (store: Store): void => {
  const fileInput: HTMLInputElement = document.createElement("input");
  fileInput.type = "file";
  fileInput.accept = "application/json";
  fileInput.addEventListener("input", () => {
    const file = fileInput.files?.[0];
    if (!file) return;

    loadFromJson(store, file).catch(e => console.error(e));
  });
  fileInput.click();
};

export const saveToJson = (store: Store): void => {
  const data: LatestSchema = {
    version: 1,
    servants: toJS(store.ownedServants),
    items: [...store.ownedItems.entries()],
  };
  save("fgo-planner.json", new Blob([JSON.stringify(data)], { type: "application/json" }));
};

export const saveToCsv = (store: Store): void => {
  const target = ({ current, target }: Target): string => `${typeof current === "number" ? current : ""},${typeof target === "number" ? target : ""}`;

  const contents = store.ownedServants.map(servant => {
    const details = store.servantLookup.get(servant.id)!;
    return (
      `${servant.id},${details.name},${target(servant.level)},${target(servant.ascension)},` +
      `${target(servant.skills[0])},${target(servant.skills[1])},${target(servant.skills[2])}`
    );
  }).join("\n");
  save("fgo-planner.csv", new Blob([contents], { type: "text/csv" }));
};
