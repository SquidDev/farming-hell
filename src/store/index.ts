import { autorun, configure, makeAutoObservable } from "mobx";
import { v4 as uuidV4 } from "uuid";

import { DataDump, Filters, GameEvent, GrailCosts, Id, IdMap, Item, OwnedServant, Priority, Servant, makeId } from "../data";
import { Requirements, addServantRequirements, newRequirements } from "../data/requirements";
import { loadFromLocalStorage, saveToLocalStorage } from "./save";
import { defaultFilters } from "./filter";
import { expId, qpId } from "../data/constants";


if (process.env.NODE_ENV === "development") {
  configure({
    enforceActions: "always",
    computedRequiresReaction: true,
    reactionRequiresObservable: true,
    observableRequiresReaction: true,
    disableErrorBoundaries: true
  });
}

const makeLookup = <T extends { id: Id<string> }>(objects: Array<T>): IdMap<T> => {
  const lookup: IdMap<T> = new Map();
  for (const object of objects) lookup.set(object.id, object);
  return lookup;
};


/**
 * The global data store of servants.
 *
 * This is powered by mobx because I can't be dealing with Redux. Imperative
 * mutations to the store a propagated globally throughout the app.
 *
 * The store holds two bits of information:
 *  - All servants currently in NA. This is not tracked by mobx (and is
 *    notionally immutable).
 *  - The list of currently owned servants.
 */
export class Store {
  public readonly servants: Array<Servant>;
  public readonly items: Array<Item>;
  public readonly servantLookup: IdMap<Servant>;
  public readonly itemLookup: IdMap<Item>;
  public readonly eventLookup: IdMap<GameEvent>;
  public readonly expGrowth: Array<number>;
  public readonly grailCosts: GrailCosts;

  public ownedServants: Array<OwnedServant> = [];
  public ownedItems: Map<Id<"item">, number> = new Map();
  public filters: Filters = defaultFilters;

  constructor({ servants, items, events, expGrowth, grailCosts }: DataDump) {
    this.servants = servants;
    this.items = items;
    this.servantLookup = makeLookup(servants);
    this.itemLookup = makeLookup(items);
    this.eventLookup = makeLookup(events);
    this.expGrowth = expGrowth;
    this.grailCosts = grailCosts;

    makeAutoObservable(this, {
      servants: false,
      items: false,
      itemLookup: false,
      servantLookup: false,
      eventLookup: false,
      expGrowth: false,
      grailCosts: false,
    });
  }

  /**
   * Get requirements for all or a subset of servants.
   * @param filter Check whether this servant's requirements should be included.
   * @returns The accumulated requirements.
   */
  public getRequirements(filter?: (servant: OwnedServant, details: Servant) => boolean): Requirements {
    const requirements = newRequirements();
    for (const servant of this.ownedServants) {
      const details = this.servantLookup.get(servant.id);
      if (!details || (filter && !filter(servant, details))) continue;
      addServantRequirements(requirements, this, servant, details);
    }

    return requirements;
  }

  /** Get all required items. */
  public get requirements(): Requirements {
    return this.getRequirements();
  }

  public get requirementItems(): Array<{ item: Item, getter: (r: Requirements) => number }> {
    const result: Array<{ item: Item, getter: (r: Requirements) => number }> = [
      { item: this.itemLookup.get(qpId)!, getter: r => r.qp },
      { item: this.itemLookup.get(expId)!, getter: r => r.expBonus },
    ];

    for (const item of this.items) {
      if (item.id === qpId || item.id === expId) continue; // These have custom getters so skip.
      result.push({ item, getter: r => r.items.get(item.id) ?? 0 });
    }

    return result;
  }
}

export const artoriaPendragon: Id<"servant"> = makeId("servant", 100100);

/** Construct a new owned servant. */
export const newServant = (id?: Id<"servant">): OwnedServant => ({
  uid: uuidV4(),
  id: id ?? artoriaPendragon,
  priority: Priority.Low,
  level: {},
  ascension: {},
  skills: [{}, {}, {}],
});

/**
 * Construct a store driven from local storage. This is automatically saved
 * when the servants change.
 */
export const localStore = (data: DataDump): Store => {
  const store = new Store(data);

  try {
    loadFromLocalStorage(store);
  } catch (e) {
    console.error("Cannot load from local storage", e);
  }

  autorun(() => {
    try {
      saveToLocalStorage(store);
    } catch (e) {
      console.error("Cannot save to local storage", e);
    }
  });

  return store;
};
