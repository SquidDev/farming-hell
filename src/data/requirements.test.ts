import { newServant } from "../store";
import { fsStore } from "../test";
import { addServantRequirements, newRequirements } from "./requirements";

test("Computes ascension materials", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.ascension = { current: 0, target: 4 };

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});

test("Ascension defaults to level 0", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.ascension.target = 4;

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});

test("Computes skill materials", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.skills[0] = { current: 1, target: 10 };

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});

test("Skills defaults to level 1", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.skills[0].target = 10;

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});

test("Computes levelling materials", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.level = { current: 1, target: 90 };

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});

test("Level defaults to 1", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.level.target = 1;

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});

test("Computes grailing materials", async () => {
  const store = await fsStore();

  const servant = newServant();
  servant.level = { current: 80, target: 100 };

  const requirements = newRequirements();
  addServantRequirements(requirements, store, servant, store.servantLookup.get(servant.id)!);
  expect(requirements).toMatchSnapshot();
});
