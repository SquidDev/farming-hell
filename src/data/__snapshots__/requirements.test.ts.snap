// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ascension defaults to level 0 1`] = `
{
  "exp": 0,
  "expBonus": 0,
  "items": Map {
    7001 => 17,
    6512 => 18,
    7101 => 17,
    6503 => 29,
    6501 => 12,
    6506 => 5,
  },
  "qp": 4400000,
}
`;

exports[`Computes ascension materials 1`] = `
{
  "exp": 0,
  "expBonus": 0,
  "items": Map {
    7001 => 17,
    6512 => 18,
    7101 => 17,
    6503 => 29,
    6501 => 12,
    6506 => 5,
  },
  "qp": 4400000,
}
`;

exports[`Computes grailing materials (80..100) 1`] = `
{
  "exp": 437,
  "expBonus": 364,
  "items": Map {
    7101 => 12,
    6501 => 8,
    6506 => 5,
    7999 => 5,
  },
  "qp": 58000000,
}
`;

exports[`Computes levelling materials 1`] = `
{
  "exp": 450,
  "expBonus": 375,
  "items": Map {
    7001 => 17,
    6512 => 18,
    7101 => 17,
    6503 => 29,
    6501 => 12,
    6506 => 5,
  },
  "qp": 4400000,
}
`;

exports[`Computes lvl 120 materials 1`] = `
{
  "exp": 15348,
  "expBonus": 12790,
  "items": Map {
    7999 => 15,
    10010001 => 300,
  },
  "qp": 205000000,
}
`;

exports[`Computes skill materials 1`] = `
{
  "exp": 0,
  "expBonus": 0,
  "items": Map {
    6001 => 17,
    6101 => 17,
    6503 => 44,
    6201 => 17,
    6512 => 36,
    6507 => 15,
    6506 => 10,
    6999 => 1,
  },
  "qp": 54400000,
}
`;

exports[`Level defaults to 1 1`] = `
{
  "exp": 0,
  "expBonus": 0,
  "items": Map {},
  "qp": 0,
}
`;

exports[`Skills defaults to level 1 1`] = `
{
  "exp": 0,
  "expBonus": 0,
  "items": Map {
    6001 => 17,
    6101 => 17,
    6503 => 44,
    6201 => 17,
    6512 => 36,
    6507 => 15,
    6506 => 10,
    6999 => 1,
  },
  "qp": 54400000,
}
`;
